Implements keep alive

devstepien opened the pull request:
Adiciona dois módulos que fazem monitoração e envio de keep alive messages. 
Esses módulos podem ser plugados em qqr parte do sistema.
- ConnectionKeeper utilizado para mandar as mensagens de keep alive
- ConnectionMonitor utilizado para monitorar essas mensagens.

devstepien committed:
Keep Alive implementation

- 10 seconds sending keep alive messages every 2 seconds.

devstepien commented on Instant-Messenger/src/client/ConnectionKeeper.cpp:
```diff
@@ -0,0 +1,34 @@
+    while (true)
+    {
+        sleep(_this->sleepTime);
+        keepAlivePacket->isKeepAlive = true;
```
dá pra ser fora do while essa atribuição

devstepien commented on Instant-Messenger/src/client/ConnectionKeeper.cpp:
```diff
@@ -0,0 +1,34 @@
+
+
+ConnectionKeeper::ConnectionKeeper(int socket) {
+    pthread_t monitoringThread;
```
mudar para connectionKeeper thread

devstepien commented on Instant-Messenger/src/client/ConnectionKeeper.cpp:
```diff
@@ -0,0 +1,34 @@
+//
+// Created by gabriel on 11/4/20.
```
ninguém liga, Clion

devstepien commented on Instant-Messenger/include/server/ConnectionMonitor.hpp:
```diff
@@ -0,0 +1,32 @@
+//
+// Created by gabriel on 11/4/20.
```
ninguém liga, Clion

devstepien reviewed (commented):
Algumas coisas que tenho que mudar mais tarde

devstepien committed:
Some points that I left opened

Signed-off-by: gabriel-inf <gabriel.stepien.eu@gmail.com>

Andy9822 commented on Instant-Messenger/src/server/server.cpp:
```diff
@@ -198,12 +200,20 @@ namespace server {
         // Also, send reference of this instance to the new thread
         args->second = this;
-
+        pthread_create(&monitoringThread, NULL, monitorConnection, (void *) args);
```
Pra cada conexão é aberta uma thread de monitoramento de conexão.
Não seria melhor ter apenas 1 thread que monitora todas as conexões?

Analisa e dá tua opinião sincera sobre se é melhor +1 thread por cliente ou 1 thread por servidor

devstepien replied:
Uma opção. Mas gostei de abrir uma thread para monitorar cada usuário pq assim a gente conta o tempo com base no momento que o usuário conectou.

devstepien replied:
Outra coisa. Abrindo uma thread pra cada a gente meio que deixa um pouco mais genérico pq a gente começa a monitorar um socket. Só precisamos do identificador do socket. Se for fazer um monitoring por uma lista de sockets teria todo mundo que for usar esse monitor vai ter que passar essa lista e chamar a rotina de encerramento de cone'xão pra cada socket que der timeout. Do jeito que tá eu tenho certeza que, quando chegar no join dessa thread `monitoringThread` eu posso numa boa encerrar a conexão pq chegamos no timeout.

Andy9822 commented on Instant-Messenger/include/util/Packet.hpp:
```diff
@@ -15,6 +15,7 @@ struct Packet
     char message[MESSAGE_MAX_SIZE];
     int clientSocket;
     time_t timestamp;
+    bool isKeepAlive = false;
```
Eu deixaria algo mais genérico, literalmente um campo `int TYPE` ou `OPT` que determine a semântica daquele Packet.
Acredito que vamos ter mais do que 2 tipos de Packet diferentes.

devstepien replied:
Hmm, vdd. Provavelmente vamos usar esse packet para as mensagens de eleição. I'll change

Andy9822 commented on Instant-Messenger/src/client/ConnectionKeeper.cpp:
```diff
@@ -0,0 +1,30 @@
+void * ConnectionKeeper::sendKeepAliveForever(void *args) {
+    ConnectionKeeper *_this = (ConnectionKeeper *) args;
+
+    Packet *keepAlivePacket = new Packet();
```
Talvez trocar isso pra um constructor polimórfico que recebe apenas o tipo de Packet.

devstepien replied:
Acho que o suficiente vai ser a gente receber o TYPE no construtor como tu tinha falando no comentário anterior e já setar num atributo type do packet. Sem mais engineering pra isso

Andy9822 commented on Instant-Messenger/include/util/definitions.hpp:
```diff
@@ -38,4 +38,12 @@
+#define CONNECTION_LIVE 10
+#endif
+
+#ifndef CONNECTION_COUNTDOWN_STEP
```
Judge me
But would'nt be better smth like `KEEP_ALIVE_INTERVAL` ?

devstepien replied:
Naming is important and you definitely got a better one for this. I'll change

Andy9822 commented on Instant-Messenger/include/util/definitions.hpp:
```diff
@@ -38,4 +38,12 @@
+#endif
+
+#ifndef CONNECTION_LIVE
+#define CONNECTION_LIVE 10
```
Like in the comment below, what about smth like `KEEP_ALIVE_TIMEOUT`?

Andy9822 commented on Instant-Messenger/src/server/ConnectionMonitor.cpp:
```diff
@@ -0,0 +1,59 @@
+        _this->socketMapSemaphore->wait();
+        _this->socketCountdown[socket] -= CONNECTION_COUNTDOWN_STEP;
+        _this->socketMapSemaphore->post();
+        connectionIsValid = _this->socketCountdown[socket] > 0;
```
Agora entendi o `CONNECTION_COUNTDOWN_STEP`

Não sei, vê tua opinião cmg: não seria mais simples apenas ir testando timestamps? 
i.e. `sleep()` e testar o `timestamp > KEEP_ALIVE_TIMEOUT` ?

devstepien replied:
Pode ser mais simples. Não me perguntei como fazer a comparação com times pq tinha essa implementação aí na cabeça. Se for algo que dê pra trocar nos próximos 30 min eu o farei.

devstepien replied:
no final vamos estar comparando se a difereça entre dois inteiros são maior do que o timeout configurado. Tem que usar uma função 'difftime(now,mktime(&old_date));' esse resultado vai ser em segundos. Eu não gostar em ter que mudar isso :cry:  , na real pra mim não faz diferença

devstepien replied:
pensando mais um pouquinho, a possibilidade de remover aquele decremento usando comparação de timestamps pode justificar, to vendo aqui

devstepien replied:
feito

Andy9822 reviewed (changes requested):
Dá olhada sobre os pontos que levantei e dá tua opinião sobre

Andy9822 commented on Instant-Messenger/src/server/server.cpp:
```diff
@@ -198,12 +200,20 @@ namespace server {
+        Server *_this = (Server *) args_pair->second;
 
+        cout << "monitorConnection" << endl;
+        _this->connectionMonitor->monitor(&client_socketfd);
```
Cuidado com o `client_socketfd`.
É um ponteiro que foi alocado memória pra ele porém em `listenClientCommunication()` se dá `free()`. 
Pode acontecer do monitor ir ler um ponteiro que já foi dado `free()`

devstepien replied:
sim, mais seguro vai ser tu dar o malloc desse cara pra manter esse endereço reservado só pra chegar safe no nosso método de monitoring e, de forma separada, dar o free nele dps

devstepien committed:
Allowing more types of packages

As default, a packet will be created as a MESSAGE_PACKET. If this is another type of packet, we use a constructor to say the type. Rip OOP.

Signed-off-by: gabriel-inf <gabriel.stepien.eu@gmail.com>

devstepien committed:
Changes ConnectionMonitor logic to use time

- Uses time as suggested in the PR.
- Enhances the naming to increases readability.

Signed-off-by: gabriel-inf <gabriel.stepien.eu@gmail.com>

devstepien committed:
Reserves another addressing space for the monitoring args

Signed-off-by: gabriel-inf <gabriel.stepien.eu@gmail.com>
